src/storage/txn/flow_controller.rs

```diff
@@ -0,0 +1,934 @@
+
+        let half = self.records.len() / 2;
+        let mut left = 0.0;
+        let mut right = 0.0;
+        for (i, r) in self.records.iter().enumerate() {
+            if i + 1 < half {
+                left += r.0 as f64;
+            } else if i + 1 > half {
+                right += r.0 as f64;
+            } else {
+                if self.records.len() % 2 == 0 {
+                    left += r.0 as f64;
+                } else {
+                    continue;
+                }
+            }
+        }
+        let elapsed = duration_to_sec(
+            self.records
+                .back()
+                .unwrap()
+                .1
+                .duration_since(self.records.front().unwrap().1),
+        );
+        (right - left) / half as f64 / (elapsed / 2.0)
+    }
+
+    pub fn trend(&self) -> Trend {
+        if self.records.len() == 0 {
+            return Trend::NoTrend;
+        } else if self.records.len() == 1 {
+            return Trend::OnlyOne;
+        }
+
+        // Calculate the average of left and right parts
+        let half = self.records.len() / 2;
+        let mut left = 0;
+        let mut right = 0;
+        for (i, r) in self.records.iter().enumerate() {
+            if i + 1 < half {
+                left += r.0;
+            } else if i + 1 > half {
+                right += r.0;
+            } else {
+                if self.records.len() % 2 == 0 {
+                    left += r.0;
+                } else {
+                    continue;
+                }
+            }
+        }
+
+        if right > left + 2 {
+            Trend::Increasing
+        } else if right < left - 2 {
+            Trend::Decreasing
+        } else {
+            Trend::NoTrend
+        }
+    }
+}
+
+// CFFlowChecker records some statistics and states related to one CF.
+// These statistics falls into five categories:
+//   * memtable
+//   * L0 files
+//   * flush flow
+//   * L0 flow (compaction read flow of L0)
+//   * pending compaction bytes
+// And all of them are collected from the hook of RocksDB's event listener.
+struct CFFlowChecker {
+    last_num_memtables: Smoother<60>,
+    memtable_debt: f64,
+    init_speed: bool,
+
+    last_num_l0_files: u64,
+    last_num_l0_files_from_flush: u64,
+    long_term_num_l0_files: Smoother<20>,
+
+    last_flush_bytes_time: Instant,
+    last_flush_bytes: u64,
+    short_term_flush_flow: Smoother<10>,
+    long_term_flush_flow: Smoother<60>,
+
+    last_l0_bytes: u64,
+    last_l0_bytes_time: Instant,
+    short_term_l0_flow: Smoother<3>,
+
+    long_term_pending_bytes: Smoother<60>,
+
+    on_start_memtable: bool,
+    on_start_l0_files: bool,
+    on_start_pending_bytes: bool,
+}
+
+impl Default for CFFlowChecker {
+    fn default() -> Self {
+        Self {
+            last_num_memtables: Smoother::default(),
+            long_term_pending_bytes: Smoother::default(),
+            long_term_num_l0_files: Smoother::default(),
+            last_num_l0_files: 0,
```

Comment:
`last_num_l0_files == long_term_num_l0_files.get_recent()`?

Reply:
I explained it in the latest comment. `last_num_l0_files` records last number of L0 files right after flush or L0 compaction, whereas `long_term_num_l0_files` only records L0 files right after L0 compaction.